Isosurface mesh refinment plugin

MFraters opened the pull request:
I have been working with @lhy11009 to transform the isotherm plugin (#295) to a more general form and get it in a state that it is ready to be merged. There are a few loose end to be tied, but it seems functional and contains several tests. 

Here is the todo list on top of my head:

- [ ] Write plugin class description.
- [ ] Add background material option as `C_b`, `c_b` and/or `Background composition`.
- [ ] Convert list_of_composiiton_names to a list to a list of used key enums, which can be used to determaine what properties need to be queried. Although It might be better to leave this for a feature pull request. So maybe just remove for now. 
- [ ] Come up with a better name than isolines?

bangerth commented:
This isn't ready for review, right?

May I suggest the name "Isosurface" instead of "Isoline", since the latter is inherently 2d? Or is this really what you mean?

MFraters commented:
> This isn't ready for review, right?

No, but is should be ready soon :)

> May I suggest the name "Isosurface" instead of "Isoline", since the latter is inherently 2d? Or is this really what you mean?

You are right. Isosurface would be better. I am fine with changing it to that. @lhy11009, what do you think?

lhy11009 commented:
I agree that 'isosurface' is a good name.

lhy11009 commented:
to be more specific, will that be 'isosurfaces'? Let's get this reviewed today. I think all three of us want to use this in the models we are working on.

MFraters commented:
@lhy11009, I did the renaming and I added a changelog. Could you take a look at it so see if you think it is ready for review now or whether you still want to change something?

The index+doc tester complained about  'fatal: ref HEAD is not a symbolic ref' during the compile stage, so I rebased. I hope that fixes it.

MFraters commented:
I am not sure why one of the github testers fails @tjhei or @gassmoeller suggestions?

tjhei commented:
This is due to the unity build. My PR contains a fix: https://github.com/geodynamics/aspect/pull/3692/commits/69c2ce7d2be57e583f2a7569a42b07011fc9a6e2

I will make a separate PR so you can rebase.

lhy11009 commented:
Menno, that's already fine with me. I think we can go for a review.

MFraters commented:
All the tests pass now and this pull request is ready for review

tjhei commented on include/aspect/mesh_refinement/isosurfaces.h:
```diff
@@ -0,0 +1,130 @@
+          std::vector<double> min_values;
+          std::vector<double> max_values;
+          double min_refinement;
+          double max_refinement;
```
Please document. Why is a refinement a double?

MFraters replied:
I have no idea... I will change it to an unsigned int.

MFraters replied:
well, that should be int, because cell->level() is an int

tjhei commented on include/aspect/mesh_refinement/isosurfaces.h:
```diff
@@ -0,0 +1,130 @@
+          std::vector<Property> properties;
+      };
+
+      unsigned int min_max_string_to_int(const std::string &string, const unsigned int minimum_refinement_level, const unsigned int  maximum_refinement_level);
```
What does this function do? Is this a detail of the plugin? Then move it into the .cc file.

tjhei commented on include/aspect/mesh_refinement/isosurfaces.h:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    /**
+     * A class that implements a Isosurfaces mesh refinement plugin. This
```
```suggestion
     * A class that implements an Isosurfaces mesh refinement plugin. This
```

tjhei commented on include/aspect/mesh_refinement/isosurfaces.h:
```diff
@@ -0,0 +1,130 @@
+    /**
+     * A class that implements a Isosurfaces mesh refinement plugin. This
+     * plugin allows for setting a minimum and maximum refinment level in
+     * a part of the model domain where a value (e.g. Tempearture) is between
```
```suggestion
     * a part of the model domain where a value (e.g. Temperature) is between
```

tjhei commented on tests/isosurfaces_simple_box_3D/screen-output:
```diff
@@ -0,0 +1,23 @@
+*** Timestep 0:  t=0 years
+
+   Postprocessing:
+     Writing graphical output: output-isosurfaces_simple_box_3D/solution/solution-00000
```
Do we need these large .gnuplot files or is it enough to see the number of DoFs to make sure things are correct?

MFraters replied:
ah, I didn't realise it was that big. I think looking at the dofs is good enough of an indicator whether something changed

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+    namespace Internal
+    {
+      bool
+      Isosurface::values_are_in_range(const std::vector<double> values) const
```
```suggestion
      Isosurface::are_all_values_in_range(const std::vector<double>& values) const
```

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+      Isosurface::values_are_in_range(const std::vector<double> values) const
+      {
+        // This assumes that all the vectors in isosurfaces already have the
+        // same length.
```
that should be in the documentation of the function.

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+          }
+        else
+          {
+            auto p = std::find(available_compositions.begin(), available_compositions.end(), property_name);
```
Hm, what if I call one of my compositional fields "background"?

MFraters replied:
I wasn't really aware of it, but @lhy11009 pointed out that this name seems to be already kind of reserved by the `parse_map_to_double_array` function. I have added an assert  in the parse parameters. I am open to other solutions. @lhy11009, what do you think?

lhy11009 replied:
Yes, we have already reserved that for phase changes. I think it would be better to make it so here.

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+          }
+      }
+
+      unsigned int min_max_string_to_int(const std::string &string_value, const unsigned int minimum_refinement_level, const unsigned int  maximum_refinement_level)
```
I don't understand what this function does. You seem to be parsing an expression here. Consider using a FunctionParser or can you write a few unittests?

MFraters replied:
I have moved it, documented it and added a unit test :)

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+                            }
+                          else if (isosurface.properties[index].name == Internal::PropertyName::Composition)
+                            {
+                              values[index] = std::min(std::max(in.composition[i_quad][isosurface.properties[index].index], 0.0), 1.0);
```
Why are compositions assumed to be between 0. and 1.?

MFraters replied:
I see no reason to restrict that here, but I assume this is based on what is done in `parse_map_to_double_array`. Is that correct @lhy11009?

lhy11009 replied:
I think @tjhei 's comments make sense. But I just copy from the 'volume_fraction' function from the MaterialUtility class. Right now I am not sure what to do. Let discuss it.

tjhei replied:
This is still an open problem.

MFraters replied:
ah, sorry, I missed it somehow. I have just removed the bounds checks.

tjhei commented on include/aspect/mesh_refinement/isosurfaces.h:
```diff
@@ -0,0 +1,130 @@
+           */
+          Property(const std::string &property_name,const std::vector<std::string> &available_compositions);
+
+          PropertyName name;
```
Also, "name" is a bit weird to use here as it is more of a type or not?

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+                              double sum_composition = 0.0;  // Compute background material fraction
+                              for (unsigned i=0 ; i < in.composition[i_quad].size() ; ++i)
+                                {
+                                  sum_composition += std::min(std::max(in.composition[i_quad][i], 0.0), 1.0);
```
You assume that all compositional fields participate in a case where they sum up to 1-background. That is quite restrictive as compositional fields are also used for other things.

MFraters replied:
same as comment above, @lhy11009 what is your opinion on this?

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+                          // level, any refinement flag should be cleared.
+                          if (cell->level() >= isosurface.max_refinement)
+                            {
+                              clear_refine = true;
```
What happens if two "isosurfaces" overlap. I assume that the "last" one wins? Is that desired?

MFraters replied:
Yes. I don't think there is a  most desired way of doing it, but it is a clear way. I have added a sentence about that in the declare_entry of isosurfaces.

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+                }
+              if (clear_refine == true)
+                {
+                  cell->clear_refine_flag ();
```
I think you would want to clear flags before setting them. Otherwise, the set_ might throw.

tjhei replied:
This is still open.

MFraters replied:
I somehow only did it for the coarsen flag. It's fixed now.

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+                  cell->clear_refine_flag ();
+                }
+            }
+          ++i;
```
what is this used for? Maybe give it a more descriptive name?

MFraters replied:
It wasn't used, so I removed it.

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+                             "A list of isosurface separated by semi-colins (;). Each isosurface entry consists of "
+                             "multiple entries separted by a comma. The first two entries indicate the minimum and maximum "
+                             "refinement levels respectively. The entries after the first two describe the field the isosurface "
+                             "applies to, followed by a colin (:) followed by the minimum and maximum grid levels seperated by "
```
```suggestion
                             "applies to, followed by a colon (:) followed by the minimum and maximum grid levels separated by "
```

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+    {
+      // lookup the minimum and maximum refinment
+      prm.enter_subsection("Mesh refinement");
+      unsigned int minimum_refinement_level = Utilities::string_to_int(prm.get ("Minimum refinement level"));
```
can you get this from this->get_parameters() instead?

MFraters replied:
done

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+      unsigned int maximum_refinement_level = Utilities::string_to_int(prm.get("Initial global refinement")) + Utilities::string_to_int(prm.get("Initial adaptive refinement"));
+      prm.leave_subsection();
+
+      const std::vector<std::string> compositions = this->introspection().get_composition_names();
```
```suggestion
      const std::vector<std::string>& compositions = this->introspection().get_composition_names();
```

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+                          ExcMessage("An isosurface needs to contain at least 3 entries, but isosurface " + std::to_string(isosurface_entry_number)
+                                     + " contains  only " +  std::to_string(field_entries.size()) + " entries: " + isosurface_entry + "."));
+
+              // convert a potential min, min+1, min + 1, min+10, max, max-1, etc. to actual integers.
```
This is not documented in the parameter description, could this be added?

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+    ASPECT_REGISTER_MESH_REFINEMENT_CRITERION(Isosurfaces,
+                                              "isosurfaces",
+                                              "A mesh refinement criterion that computes "
+                                              "refinement indicators between two iso-surfaces of"
```
```suggestion
                                              "refinement indicators between two iso-surfaces of "
```

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+                                              "\n\n"
+                                              "The way these indicators are derived on each isosurface is by "
+                                              "checking the conditions whether solutions of specific "
+                                              "fields are within the ranges of values given. If these conditions"
```
```suggestion
                                              "fields are within the ranges of values given. If these conditions "
```

tjhei replied:
and all the other lines below.

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+                                              "fields are within the ranges of values given. If these conditions"
+                                              "hold, then an indicator is either put on or taken off"
+                                              "in order to secure mesh refinement within the range of levels"
+                                              "given. Usage of this plugin allows user to put an conditional"
```
```suggestion
                                              "given. Usage of this plugin allows user to put a conditional"
```

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,391 @@
+                             "refinement levels respectively. The entries after the first two describe the field the isosurface "
+                             "applies to, followed by a colin (:) followed by the minimum and maximum grid levels seperated by "
+                             "bar (|). An example for an isosurface is '0, 2, Temperature: 300 | 600'; 2, 2, C\\_1: 0.5 | 1");
+
```
You do not document "min", "max", or "background" here.

tjhei commented:
I jumped into the code without knowing anything about what it does, and it was challenging to understand it by reading the code. What you are trying to do is quite complicated. Sorry for some of the "dumb" questions.
I would like us to think about the grammar for the parameters to make things a bit easier for the user.

MFraters commented:
Thanks for the extensive review, it is really appreciated! I made a quick scan through your comments and I will try to address them later today.

MFraters commented on include/aspect/mesh_refinement/isosurfaces.h:
```diff
@@ -0,0 +1,130 @@
+        Background
+      };
+
+      class Property
```
I looked it up and it looks very similar indeed. If this would be the final state of this plugin we might be able to use that instead, even through it also does other things. But the idea for this plugin is that we want to add other properties like viscosity, density, stress etc in the future as well. So I think having a specialized structure make more sense to me.

MFraters commented on include/aspect/mesh_refinement/isosurfaces.h:
```diff
@@ -0,0 +1,130 @@
+           * Checks whether the provided values are between the min and max value  for
+           * those property of the isosurface. This function assumes that the order of the
+           * provided values matches the order in which the properties are stored.
+           */
```
The length of the vector is dependent on the user input. There is one entry for each property type in a single isosurface.

MFraters commented:
I had a discussion with @lhy11009 and @naliboff this morning, and we decided that dealing with the background composition ouside of the material model correctly is actually more difficult than we hoped for, and will require some other to the material model. Since we want this plugin to be merged as soon as possible, we decided that for now it is best to just remove it. I can be added in back later.

I think this pull request is ready for a final review/to be merged.

tjhei commented:
> nd we decided that dealing with the background composition ouside of the material model correctly is actually more difficult than we hoped for, and will require some other to the material model.

I would suggest that we think about a system that allows the user to specify in the .prm that the following n-1 compositional fields plus 1.0-sum(them) form different species. This could then be used in many places to query information. Maybe "background" could also be a compositional field that is set to 1.0-sum(others) automatically.

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,387 @@
+                                              "fields are within the ranges of values given. If these conditions "
+                                              "hold, then an indicator is either put on or taken off "
+                                              "in order to secure mesh refinement within the range of levels "
+                                              "given. Usage of this plugin allows user to put a conditional "
```
```suggestion
                                              "given. Usage of this plugin allows the user to put a conditional "
```

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,387 @@
+                                              "isosurface entry consists of multiple entries separated by a comma. "
+                                              "The first two entries indicate the minimum and maximum refinement "
+                                              "levels respectively. The entries after the first two describe the "
+                                              "field the isosurface applies to, followed by a colin (:) followed by "
```
```suggestion
                                              "field the isosurface applies to, followed by a colon (:) followed by "
```

tjhei commented on tests/isosurfaces_simple_box_2D.prm:
```diff
@@ -0,0 +1,106 @@
+  set Strategy = isosurfaces
+  subsection Isosurfaces
+    set Isosurfaces = 4,    4,    Temperature: 0 | 600,   C_2 : 0.0 | 0.5 ;\
+                      0,    0,    Temperature: 1600 | 3000,   C_2 : 0.0 | 0.5
```
would you mind changing one of them to only use one field?

MFraters replied:
sure, I did that for the `isosurfaces_simple_box_2D_min_max_2` test

MFraters commented:
I think that I have addressed all your comments. I have lowered the resolution of the 2D tests and purged their gnuplot results from the history and added them back in the end. I think I did that correctly, and the tests still work (including the extra tests), but it might be good to do a quick recheck of the pull request in case I made a mistake and missed it.

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,398 @@
+                             "applies to, followed by a colon (:), which again followed by the minimum and maximum grid levels separated by "
+                             "bar (|). An example for an isosurface is '0, 2, Temperature: 300 | 600; 2, 2, C\\_1: 0.5 | 1'. "
+                             "In this example the mesh refinement is kept between level 0 and level 2 if the temperature is between "
+                             "300 and 600 and the compositional field C\\_1 is between 0.5 and 1. "
```
that is not what I understand the line above does (C_1 is kept at 2).

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,398 @@
+                             "\n\n"
+                             "The first two entries for each isosurface, describing the minimum and maximum grid levels, can be "
+                             "two numbers or contain one of the key values 'min' and 'max'. This indicates the key will be replaced with "
+                             "the global minimum and maximum refinement levels. The 'min' and 'max' keys will also allow "
```
```suggestion
                             "the global minimum and maximum refinement levels. The 'min' and 'max' keys also allows "
```

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,398 @@
+                             "two numbers or contain one of the key values 'min' and 'max'. This indicates the key will be replaced with "
+                             "the global minimum and maximum refinement levels. The 'min' and 'max' keys will also allow "
+                             "values to be added to or substracted from them. This is done by adding a '+' or '-' and "
+                             "number behind them (e.g. min+2 or max-1). Note that you can't substract a value from a minimum value or "
```
```suggestion
                             "a number behind them (e.g. min+2 or max-1). Note that you can't substract a value from a minimum value or "
```

MFraters commented:
I messed up with git a bit, will fix it.

MFraters commented:
Oke, I think I fixed it.

MFraters commented:
I am sorry it took this long, but I think it is ready now for a final review.

MFraters commented:
@tjhei Could you take a look at this? I want to use this soon, and it would be nice if it was merged. Thanks!

tjhei reviewed (commented):
I am sorry for the long delay. I have a couple of minor requests.

MFraters commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,399 @@
+              }
+            else
+              {
+                AssertThrow(string.compare(0,4,"min-") != 0,
```
Yes, you are right, only `min+` plus a number or `min` should be allowed

MFraters commented on include/aspect/mesh_refinement/isosurfaces.h:
```diff
@@ -0,0 +1,146 @@
+        Composition
+      };
+
+      struct Property
```
Yes, currently is functions very similar to the `Simulator::AdvectionField` with a few extra convenient functions. I did not realize this and looked into it, but I don't think it can be replaced, because the plan is for the property to also contain other things than advection fields such as density or strain-rate.

MFraters commented:
Thanks for the through review Timo. I addressed your comments and changed one more thing. I changed the following lines from 

```C++
if(coarsen == true)
  clear_coarsen == false
```
to

https://github.com/geodynamics/aspect/blob/292397ed354d302847cb2e388eaf3f886bedeb18/source/mesh_refinement/isosurfaces.cc#L240-L241

because I noticed that the old code was in some cases alternating refining and coarsening cells on the border of a isosurface each timestep. Now it always chooses the higher refinement option, as the comment above those lines states.

MFraters commented:
@tjhei could you take a look at this? It would be nice if it can this be merged before the next release. Thanks!

tjhei commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+      {
+        prm.enter_subsection("Isosurfaces");
+        {
+          prm.declare_entry ("Isosurfaces", "depth",
```
"depth" does not seem to be  sensible default. Maybe leave it empty?

MFraters replied:
good point, I will just leave it empty.

tjhei reviewed (approved):
Please fix the last comment and rebase/squash your commits a bit. Otherwise good to go. Sorry for this lengthy process.

tjhei commented:
Finally, it might make sense if somebody else @geodynamics/aspect-developers takes a quick look.

MFraters commented:
@tjhei Thanks for the quick reply and all the reviewing!

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+                             "A list of isosurfaces separated by semi-colons (;). Each isosurface entry consists of "
+                             "multiple entries separated by a comma. The first two entries indicate the minimum and maximum "
+                             "refinement levels respectively. The entries after the first two describe the fields the isosurface "
+                             "applies to, followed by a colon (:), which again followed by the minimum and maximum grid levels separated by "
```
```suggestion
                             "applies to, followed by a colon (:), which is again followed by the minimum and maximum property values separated by "
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+                             "bar (|). An example for an isosurface is '0, 2, Temperature: 300 | 600; 2, 2, C\\_1: 0.5 | 1'. "
+                             "In this example the mesh refinement is kept between level 0 and level 2 if the temperature is between "
+                             "300 and 600 and at level 2 when the compositional field C\\_1 is between 0.5 and 1. If both happen at "
+                             "the same location, the last one is set, which in this case would be a fixed level 2."
```
```suggestion
                             "the same location, the conditions for the last isosurface are used, which in this case would be a fixed level 2."
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+                             "maximum, it will simply use the global minimum and maximum values respectively. The same holds for any "
+                             "mesh refinement level below the global minimum or above the global maximum."
+                             "\n\n"
+                             "If isosurfaces overlap, the last one sets the values.");
```
```suggestion
                             "If isosurfaces overlap, the last isosurface listed in the input determines the refinement levels.");
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+      {
+        // lookup the minimum and maximum refinement
+        const unsigned int minimum_refinement_level = this->get_parameters().min_grid_level;
+        const unsigned int maximum_refinement_level = this->get_parameters().initial_global_refinement + this->get_parameters().initial_adaptive_refinement;
```
```suggestion
        const unsigned int maximum_refinement_level = this->get_parameters().initial_global_refinement + 
                                                      this->get_parameters().initial_adaptive_refinement;
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+          unsigned int isosurface_entry_number = 0;
+          for (auto &isosurface_entry : isosurface_entries)
+            {
+              ++isosurface_entry_number;
```
Why not start at 0?

anne-glerum replied:
Sure, fine with me :)

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+                  properties.push_back(internal::Property(key_and_value[0], compositions)); // convert key to property name
+                  const std::vector<std::string> values = dealii::Utilities::split_string_list(key_and_value[1], '|');
+                  AssertThrow(values.size() == 2,
+                              ExcMessage("Both a maximum and a minimum values are required for each isosurface."));
```
```suggestion
                              ExcMessage("Both a maximum and a minimum value are required for each isosurface."));
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+                  min_value_inputs.push_back(Utilities::string_to_double(values[0]));  // get min and max values of the range
+                  max_value_inputs.push_back(Utilities::string_to_double(values[1]));
+                  AssertThrow(min_value_inputs.back() < max_value_inputs.back(),
+                              ExcMessage("The provided maximum refinement level has to be larger than the provided minimum refinement level."));
```
```suggestion
                              ExcMessage("The provided maximum property value has to be larger than the provided minimum property value."));
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+    ASPECT_REGISTER_MESH_REFINEMENT_CRITERION(Isosurfaces,
+                                              "isosurfaces",
+                                              "A mesh refinement criterion that computes "
+                                              "refinement indicators between two iso-surfaces of "
```
```suggestion
                                              "coarsening and refinement indicators between two isosurfaces of "
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+                                              "refinement indicators between two iso-surfaces of "
+                                              "specific field entries (e.g. temperature, composition)."
+                                              "\n\n"
+                                              "The way these indicators are derived on each isosurface is by "
```
```suggestion
                                              "The way these indicators are derived between pairs of isosurfaces is by "
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+                                              "\n\n"
+                                              "The way these indicators are derived on each isosurface is by "
+                                              "checking whether the solutions of specific "
+                                              "fields are within the ranges of values given. If these conditions "
```
```suggestion
                                              "fields are within the ranges of the isosurface values given. If these conditions "
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+                                              "The way these indicators are derived on each isosurface is by "
+                                              "checking whether the solutions of specific "
+                                              "fields are within the ranges of values given. If these conditions "
+                                              "hold, then an indicator is either put on or taken off "
```
```suggestion
                                              "hold, then coarsening and refinement indicator are set such that "
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+                                              "checking whether the solutions of specific "
+                                              "fields are within the ranges of values given. If these conditions "
+                                              "hold, then an indicator is either put on or taken off "
+                                              "in order to secure mesh refinement within the range of levels "
```
```suggestion
                                              "the mesh refinement levels lies within the range of levels "
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+                                              "of the compositional fields which are either specified by user or set up "
+                                              "as C\\_0, C\\_1, etc."
+                                              "\n\n"
+                                              "Usage: A list of isosurface separated by semi-colons (;). Each "
```
```suggestion
                                              "Usage: A list of isosurfaces separated by semi-colons (;). Each "
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+                                              "The first two entries indicate the minimum and maximum refinement "
+                                              "levels respectively. The entries after the first two describe the "
+                                              "fields the isosurface applies to, followed by a colon (:), which again "
+                                              "followed by the minimum and maximum grid levels separated by bar (|). An "
```
```suggestion
                                              "is followed by the minimum and maximum field values separated by bar (|). An "
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+                                              "levels respectively. The entries after the first two describe the "
+                                              "fields the isosurface applies to, followed by a colon (:), which again "
+                                              "followed by the minimum and maximum grid levels separated by bar (|). An "
+                                              "example for an isosurface is '0, 2, Temperature: 300 | 600; 2, 2, C\\_1: 0.5 | 1'. "
```
```suggestion
                                              "example for two isosurface entires is '0, 2, Temperature: 300 | 600; 2, 2, C\\_1: 0.5 | 1'. "
```

anne-glerum replied:
Please also add the example where two fields are set for one set of min/max grid levels (like in the tests).

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+                                              "fields the isosurface applies to, followed by a colon (:), which again "
+                                              "followed by the minimum and maximum grid levels separated by bar (|). An "
+                                              "example for an isosurface is '0, 2, Temperature: 300 | 600; 2, 2, C\\_1: 0.5 | 1'. "
+                                              "If isosurfaces overlap, the last one sets the values. "
```
```suggestion
                                              "If isosurfaces overlap, the last isosurface entry specified in the input file dictates refinement. "
```

anne-glerum replied:
It would also be good to add how this plugin interacts with other refinement strategies. It will override strategies like viscosity or temperature, but how does it interact with `minimum refinement function` for example? Does the order in which the tag_additional_cells strategies are evaluated matter then?

anne-glerum commented on tests/isosurfaces_simple_box_2D.prm:
```diff
@@ -0,0 +1,105 @@
+# This input file tests the isosurface mesh refinment plugin
```
```suggestion
# This input file tests the isosurface mesh refinement plugin
```

anne-glerum commented on tests/isosurfaces_simple_box_2D.prm:
```diff
@@ -0,0 +1,105 @@
+end
+
+
+# We take a box geometry and increase the repetitions
```
This comment does not apply. Please rm

anne-glerum commented on tests/isosurfaces_simple_box_2D.prm:
```diff
@@ -0,0 +1,105 @@
+  set Initial adaptive refinement        = 2
+  set Strategy = isosurfaces
+  subsection Isosurfaces
+    set Isosurfaces = 4,    4,    Temperature: 0 | 600,   C_2 : 0.0 | 0.5 ;\
```
Maybe you can add a comment here that although the isosurface entry specifies min/max level 4, refinement level will never be more than 2 because initial global + initial adaptive = 2

anne-glerum commented on tests/isosurfaces_simple_box_2D_custom_fieldnames.prm:
```diff
@@ -0,0 +1,17 @@
+# This input file tests the isosurface mesh refinment plugin
```
```suggestion
# This input file tests the isosurface mesh refinement plugin
```

anne-glerum commented on tests/isosurfaces_simple_box_2D_custom_fieldnames.prm:
```diff
@@ -0,0 +1,17 @@
+  set Initial adaptive refinement        = 2
+  set Strategy = isosurfaces
+  subsection Isosurfaces
+    set Isosurfaces = 4,    4,    Temperature: 0 | 600,   FieldB : 0.0 | 0.5 ;\
```
Maybe you can add a comment here that although the isosurface entry specifies min/max level 4, refinement level will never be more than 2 because initial global + initial adaptive = 2

anne-glerum commented on tests/isosurfaces_simple_box_2D_custom_fieldnames_fail.prm:
```diff
@@ -0,0 +1,19 @@
+# This input file tests the isosurface mesh refinment plugin
```
```suggestion
# This input file tests the isosurface mesh refinement plugin
```

anne-glerum commented on tests/isosurfaces_simple_box_2D_min_max_1.prm:
```diff
@@ -0,0 +1,12 @@
+# This input file tests the isosurface mesh refinment plugin
```
```suggestion
# This input file tests the isosurface mesh refinement plugin
```

anne-glerum commented on tests/isosurfaces_simple_box_2D_min_max_2.prm:
```diff
@@ -0,0 +1,12 @@
+# This input file tests the isosurface mesh refinment plugin
```
```suggestion
# This input file tests the isosurface mesh refinement plugin
```

anne-glerum commented on tests/isosurfaces_simple_box_3D.prm:
```diff
@@ -0,0 +1,62 @@
+# This input file tests the isosurface mesh refinment plugin
```
```suggestion
# This input file tests the isosurface mesh refinement plugin
```

anne-glerum commented on tests/isosurfaces_simple_box_3D.prm:
```diff
@@ -0,0 +1,62 @@
+
+set Dimension = 3
+
+# We take a box geometry and increase the repetitions
```
These comments do not apply, please rm. Repetitions can also be removed, default is 1

anne-glerum replied:
Still open. The comment about the repetitions does not apply and repetitions of 1 can be removed. I know it doesn't matter for the function of this test, it's just cleaner.

anne-glerum commented on tests/isosurfaces_slab_3D.prm:
```diff
@@ -0,0 +1,62 @@
+# This input file tests the box geometry in
+# combination with the isosurfaces plugin to refine a 2D slab
```
```suggestion
# combination with the isosurfaces plugin to refine a 3D slab
```

anne-glerum reviewed (changes requested):
Hi @MFraters , great plugin. I have some small things to address, and then this can be merged!

MFraters commented on include/aspect/mesh_refinement/isosurfaces.h:
```diff
@@ -0,0 +1,146 @@
+/*
+  Copyright (C) 2014 - 2019 by the authors of the ASPECT code.
```
I think it is usually done automatically every release with a script, but it won't hurt to update it now :)

MFraters commented on include/aspect/mesh_refinement/isosurfaces.h:
```diff
@@ -0,0 +1,146 @@
+
+          std::vector<double> min_values;
+          std::vector<double> max_values;
+          int min_refinement;
```
Yes, it could be, but the cell->level() is an int (I don't know why), so it will give warnings. I could cast it to an int every time it is compared with the cell level, but I think this is the more clean solution.

MFraters commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+                }
+
+              // if both coarsen and refine are true, give preference to refinement
+              if (coarsen == true && refine == true)
```
There can be multiple isosurfaces, where one may say coarsen and the other refine. I have rewritten the plugin declare entry part on this.

MFraters commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,401 @@
+
+      /**
+       * Convert a string describing the refinement level which potentially contains a min
+       * or max statement included in them to an integer. The format is "min", "max", "min+X", "max-X", or "X", where X is a positive integer.
```
rephrased

MFraters commented on include/aspect/mesh_refinement/isosurfaces.h:
```diff
@@ -0,0 +1,146 @@
+    /**
+     * A class that implements an Isosurfaces mesh refinement plugin. This
+     * plugin allows for setting a minimum and a maximum refinement levels in
+     * a part of the model domain where a value (e.g. Temperature) is between
```
I have rephrased it.

MFraters commented:
Thanks for the thorough review Anne! I have addressed your comments, and tried to answer your questions. Let me know if you have want further changes or have other questions.

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,404 @@
+                                              "The way these indicators are derived between pairs of isosurfaces is by "
+                                              "checking whether the solutions of specific "
+                                              "fields are within the ranges of the isosurface values given. If these conditions "
+                                              "hold, then coarsening and refinement indicator are set such that "
```
```suggestion
                                              "hold, then coarsening and refinement indicators are set such that "
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,404 @@
+                                              "The first two entries indicate the minimum and maximum refinement "
+                                              "levels respectively. The entries after the first two describe the "
+                                              "fields the isosurface applies to, followed by a colon (:), which again "
+                                              "is followed by the minimum and maximum field values separated by bar (|). An "
```
```suggestion
                                              "is followed by the minimum and maximum field values separated by a bar (|). An "
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,404 @@
+                                              "fields the isosurface applies to, followed by a colon (:), which again "
+                                              "is followed by the minimum and maximum field values separated by bar (|). An "
+                                              "example for two isosurface entires is '0, 2, Temperature: 300 | 600; 2, 2, C\\_1: 0.5 | 1'. "
+                                              "If both isoterm entires are triggered at the same location and the current refinement level is 1, "
```
```suggestion
                                              "If both isoterm entries are triggered at the same location and the current refinement level is 1, "
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,404 @@
+                                              "is followed by the minimum and maximum field values separated by bar (|). An "
+                                              "example for two isosurface entires is '0, 2, Temperature: 300 | 600; 2, 2, C\\_1: 0.5 | 1'. "
+                                              "If both isoterm entires are triggered at the same location and the current refinement level is 1, "
+                                              " it means that the first isoline will not set any flag and the second isoline will set a refinement flag. "
```
```suggestion
                                              "it means that the first isoline will not set any flag and the second isoline will set a refinement flag. "
```

anne-glerum commented on source/mesh_refinement/isosurfaces.cc:
```diff
@@ -0,0 +1,404 @@
+                                              "This means the the cell will refine. If both the coarsening and refinement flags "
+                                              "are set, preference is given to refinement. "
+                                              "\n\n"
+                                              "The minimum and maximum refinements levels per isosurface can be provided in absolute values are relative to "
```
```suggestion
                                              "The minimum and maximum refinement levels per isosurface can be provided in absolute values relative to "
```

anne-glerum reviewed (commented):
Awesome. Just some typos left as indicated and in the name of the unit test: unit_tests/mesh_refinment -> unit_tests/mesh_refinement.

MFraters commented:
Thanks for the quick follow up review! I have addressed your comments :)

anne-glerum reviewed (approved):
Great, thanks! @tjhei do you want to do the honors?

MFraters commented:
Thanks for all the reviews over the years and I am very happy what is has become and also that this is now finally merged! :tada:  I will now close the current oldest pull request on which this pull request was based on (#295), which was opened April 2015. #861 now has the honor of being the oldest open pull request ;)
